### ISSUE TITLE
Orders with successful payment in Stripe are cancelled in Sylius

### ISSUE THREAD

**Issue by sabinebaer** [FLUX-SE/SyliusPayumStripePlugin#27]
We are using Sylius 1.5.2 in combination with Plugin version 1.2.1.

We have the issue, that the payment process in Stripe is completed successfully (the money is transferred) but in Sylius the corresponding payment is cancelled. Unfortunately we cannot reproduce the behavior but only have some cases in our production system where this happened. We took a deeper look into the logs to narrow down the problem and found out, that somewhere after the payment a request to `/order/after-pay?payum_token=xxx` is done. This request leads to status code 302 which is fine I guess.  In the cases where the strange behavior described above occurs there is another request to the exact same route -  this requests returns status code 500 with the following error:

```
FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Symfony\Component\HttpKernel\Exception\NotFoundHttpException: A token with hash `xxx` could not be found. in /sylius/vendor/payum/payum/src/Payum/Core/Bridge/Symfony/Security/HttpRequestVerifier.php:49
Stack trace:
#0 /sylius/vendor/payum/payum-bundle/Controller/CaptureController.php(39): Payum\Core\Bridge\Symfony\Security\HttpRequestVerifier->verify(Object(Symfony\Component\HttpFoundation\Request))
#1 /sylius/vendor/symfony/http-kernel/HttpKernel.php(151): Payum\Bundle\PayumBundle\Controller\CaptureController->doAction(Object(Symfony\Component\HttpFoundation\Request))
#2 /sylius/vendor/symfony/http-kernel/HttpKernel.php(68): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#3 /sylius/vendor/symfony/http-kernel/Kernel.php(198): Symfony\Compon" while reading response header from upstream, client: xx.xx.xx.xx, server: xxx, request: "GET /payment/capture/xxx HTTP/1.1"
```

It seems to be pretty clear that this error occurs for the second request as the token is invalidated by the first request. After that another request to `/payment-failed?tokenValue=xxx` is made which I guess cancels the payment.

So the question is why two requests to `/order/after-pay?payum_token=xxx` are done. Is this maybe a known problem or do you have any idea what the problem could be? We would gratefully appreciate every hint that could help us to find out if this is a problem with this plugin or maybe with Stripe itself. Thanks!

**Comment by Prometee** [comment]
@sabinebaer Hello !

I will try to help you, but first I need to know which version of this plugin and of Sylius are you using.

Meanwhile here is some explanation about how tokens are used throught `Payum` and this Sylius plugin.

`Payum` is using `Token`s to ensure an action is made only one time.
During the capture phase (when you `complete` the checkout) `Payum` (here `payum/payum-bundle`) will generate a capture `Token` which is in fact a generation of 2 `Token`s, here is an example of what it's generated in the database :

| hash      | details                     | after_url                                                     | target_url                                  | gateway_name |
|-----------|-----------------------------|---------------------------------------------------------------|---------------------------------------------|--------------|
| Y7i...UAs | [Identity class serialized] | null | https://localhost/fr_FR/order/after-pay?payum_token=Y7i...UAs  | stripe       |
| 6hX...ENo | [Identity class serialized] | https://localhost/fr_FR/order/after-pay?payum_token=Y7i...UAs | https://localhost/payment/capture/6hX...ENo | stripe       |
|           |                             |                                                               |                                             |              |

The second `Token` `target_url` will be used to redirect the visitor to start a capture after the checkout.
The second `Token` `after_url` will be the `success_url` and the `cancel_url` for Stripe.
The first `Token` will be consumed when the visitor will get back from Stripe through the `success_url` or the `cancel_url`.

To sum up, the Stripe payment flow is the following one :
1. Complete the checkout
2. Token#2 -> `target_url`
3. a page displaying what Stripe needs is displayed and a redirect is made by Stripe JS script (this page is displayed for a brief moment)
4. the Stripe Checkout Session portal is displayed to the visitor
5. Token#2 -> `after_url` is called to make the visitor come back to our website

--

What you are describing seems to be impossible because the visitor being on the Stripe portal payment page will click first on the back button and then click on "Pay" nearly at the same time.
Another thing that surprise me is the url you describe : `/payment-failed?tokenValue=xxx` this url is not part of Sylius, do you have some custom code here handling only failed payments ?

**Comment by sabinebaer** [comment]
@Prometee Thanks a lot for your answer and your detailed description of the payment flow. We are using Sylius 1.5.2 in combination with plugin version 1.2.1. We will do an update of both soon but as we have a lot of dependencies it is not possible to do it right away.

Yes, it seems to be impossible but in the access log we can see two requests with the same token in the same second:
 
```
xx.xx.xx.xx - - [07/Aug/2021:14:24:08 +0200] "GET /order/after-pay?payum_token=xxx HTTP/1.1" 302 644 "-" "Mozilla/5.0 (Linux; Android 7.0; SM-G930V Build/NRD90M) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.125 Mobile Safari/537.36 (compatible; Google-Read-Aloud; +https://support.google.com/webmasters/answer/1061943)"
xx.xx.xx.xx - - [07/Aug/2021:14:24:08 +0200] "GET /order/after-pay?payum_token=xxx HTTP/1.1" 500 5 "-" "Mozilla/5.0 (Linux; Android 7.0; SM-G930V Build/NRD90M) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.125 Mobile Safari/537.36 (compatible; Google-Read-Aloud; +https://support.google.com/webmasters/answer/1061943)"
```

In comparison a successful payment and order processing looks like the following:

```
xx.xx.xx.xx - - [11/Aug/2021:07:57:40 +0200] "GET /order/after-pay?payum_token=xxx HTTP/2.0" 302 420 "https://checkout.stripe.com/pay/cs_live_axxx" "Mozilla/5.0 (iPhone; CPU iPhone OS 14_7_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Mobile/15E148 Safari/604.1"
```

There are two differences: `HTTP/2.0` vs. `HTTP/1.1` and the referer being empty in the cases where it does not work correctly. But that could also be pure coincidence.

We have some custom handling of payment errors afterwards (`/payment-failed?tokenValue=xxx`) but as far as I can tell now the problem occurs before our handling. Thanks for your hint with the webhooks - I will check them.

So as far as I understand your explanation Stripe directly calls the after-pay route and therefore it is not a problem of your plugin. So I will also get in touch with the Stripe support and let you now if I get any information from them. However, I still gratefully appreciate every hint from your side.

**Comment by Prometee** [comment]
There is a little difference with the `v1.2.1` version of this plugin, the `after_url` wasn't used, but instead the `target_url` was was given to Stripe. But the result is the same.

You can update the plugin to the `^2` version, because the behaviour of `Payum` never change since Sylius 1.5 version. But check the webhooks settings and test it, it really important with the v1.2.1.

**Comment by gonzalovilaseca** [comment]
@sabinebaer Having the same issue here, did you ever found the reason?
